File: Web/packages/dom/test/screen-metrics.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";

import { screenMetrics } from "../src/boot.ts";

test("screen metrics use the canonical DSX breakpoint boundary table", () => {
  const cases = [
    { width: 0, height: 1, breakpoint: "sm", sizeClass: "compact", orientation: "portrait" },
    { width: 479, height: 800, breakpoint: "sm", sizeClass: "compact", orientation: "portrait" },
    { width: 480, height: 800, breakpoint: "md", sizeClass: "compact", orientation: "portrait" },
    { width: 767, height: 800, breakpoint: "md", sizeClass: "compact", orientation: "portrait" },
    { width: 768, height: 1024, breakpoint: "lg", sizeClass: "regular", orientation: "portrait" },
    { width: 1023, height: 768, breakpoint: "lg", sizeClass: "regular", orientation: "landscape" },
    { width: 1024, height: 1366, breakpoint: "xl", sizeClass: "regular", orientation: "portrait" },
  ] as const;

  for (const expected of cases) {
    assert.deepEqual(screenMetrics(expected.width, expected.height), expected);
  }
});

test("a square viewport is landscape on every renderer", () => {
  assert.deepEqual(screenMetrics(768, 768), {
    width: 768,
    height: 768,
    breakpoint: "lg",
    sizeClass: "regular",
    orientation: "landscape",
  });
});

File: Engine/Android/core/src/test/kotlin/despia/engine/ScreenMetricsTest.kt
package despia.engine

import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

/** The pure screen-plane table (`dsx.screen.*`) — moved up from the desktop host so the
 *  app renderer (:render RouterHost) and :desktop publish through one :core-tested math.
 *  Boundaries mirror the web contract (packages/dom/src/boot.ts screenMetrics). */
class ScreenMetricsTest {
    @Test
    fun canonicalBreakpointBoundariesAndSizeClassesMatchTheWebContract() {
        val expected = listOf(
            320 to ("sm" to "compact"),
            479 to ("sm" to "compact"),
            480 to ("md" to "compact"),
            767 to ("md" to "compact"),
            768 to ("lg" to "regular"),
            1023 to ("lg" to "regular"),
            1024 to ("xl" to "regular"),
        )
        expected.forEach { (width, names) ->
            val metrics = requireNotNull(screenMetrics(width, 900, 1f))
            assertEquals(names.first, metrics.breakpoint, "width=$width")
            assertEquals(names.second, metrics.sizeClass, "width=$width")
        }
    }

    @Test
    fun metricsUseDensityIndependentWindowUnitsAndStableOrientation() {
        val landscape = requireNotNull(screenMetrics(2_048, 1_440, 2f))
        assertEquals(1024.0, landscape.width)
        assertEquals(720.0, landscape.height)
        assertEquals("landscape", landscape.orientation)
        assertEquals("xl", landscape.breakpoint)

        val square = requireNotNull(screenMetrics(800, 800, 1f))
        assertEquals("landscape", square.orientation)
        assertNull(screenMetrics(0, 800, 1f))
        assertNull(screenMetrics(800, -1, 1f))
        assertNull(screenMetrics(800, 800, Float.NaN))
        assertNull(screenMetrics(800, 800, 0f))
    }

    @Test
    fun publishingOverlaysMetricsWithoutErasingLifecycleState() {
        val metrics = requireNotNull(screenMetrics(768, 1_024, 1f))
        val state = screenState(
            mapOf<Any, Any?>("phase" to "ready", "width" to -1, 7 to "ignored"),
            metrics,
        )
        assertEquals("ready", state["phase"])
        assertEquals(768.0, state["width"])
        assertEquals("portrait", state["orientation"])
        assertEquals(null, state["7"])
    }

    @Test
    fun rotationSwapsOrientationAndKeepsTheOverlayLive() {
        val portrait = requireNotNull(screenMetrics(1_080, 2_400, 3f))
        assertEquals("portrait", portrait.orientation)
        val rotated = requireNotNull(screenMetrics(2_400, 1_080, 3f))
        assertEquals("landscape", rotated.orientation)
        val state = screenState(screenState(mapOf("phase" to "ready"), portrait), rotated)
        assertEquals("ready", state["phase"])
        assertEquals(800.0, state["width"])
        assertEquals("landscape", state["orientation"])
        assertEquals("lg", state["breakpoint"])
    }
}
